src/cdk-experimental/menu/menu-stack.ts

```diff
@@ -0,0 +1,89 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {Subject, Observable} from 'rxjs';
+
+/** Events to emit as specified by the caller once the MenuStack is empty. */
+export type FocusNext = 'nextItem' | 'previousItem' | 'currentItem' | undefined;
+
+/**
+ * Interface for the elements tracked in the MenuStack.
+ */
+export interface MenuStackItem {
+  /** A reference to the previous Menus MenuStack instance. */
+  _menuStack: MenuStack;
+}
+
+/**
+ * MenuStack allows subscribers to listen for close events (when a MenuStackItem is popped off
+ * of the stack) in order to perform closing actions. Upon the MenuStack being empty it emits
+ * from the `empty` observable specifying the next focus action which the listener should perform
+ * as requested by the closer.
+ */
+export class MenuStack {
+  /** All MenuStackItems tracked by this MenuStack. */
+  private readonly _elements: MenuStackItem[] = [];
+
+  /** Emits the element which was popped off of the stack when requested by a closer. */
+  private readonly _close: Subject<MenuStackItem> = new Subject();
+
+  /** Emits once the MenuStack has become empty after popping off elements. */
+  private readonly _empty: Subject<FocusNext> = new Subject();
+
+  /** Observable which emits the MenuStackItem which has been requested to close. */
+  close(): Observable<MenuStackItem> {
+    return this._close.asObservable();
+  }
+
+  /**
+   * Observable which emits when the MenuStack is empty after popping off the last element. It
+   * emits a FocusNext event which specifies the action the closer has requested the listener
+   * perform.
+   */
+  empty(): Observable<FocusNext> {
+    return this._empty.asObservable();
+  }
+
+  /** @param menu the MenuStackItem to put on the stack. */
+  push(menu: MenuStackItem) {
+    this._elements.push(menu);
+  }
+
+  /**
+   *  Pop off the top most MenuStackItem and emit it on the close observable.
+   *  @param focusNext the event to emit on the `empty` observable if the method call resulted in an
+   *  empty stack. Does not emit if the stack was initially empty.
+   */
+  closeLatest(focusNext?: FocusNext) {
+    const menuStackItem = this._elements.pop();
+    if (menuStackItem) {
+      this._close.next(menuStackItem);
+      if (this._elements.length === 0) {
+        this._empty.next(focusNext);
+      }
+    }
+  }
+
+  /**
+   * Pop off all MenuStackItems and emit each one on the `close` observable one by one.
+   * @param focusNext the event to emit on the `empty` observable once the stack is emptied. Does
+   * not emit if the stack was initially empty.
+   */
+  closeAll(focusNext?: FocusNext) {
+    if (this._elements.length) {
```

Comment:
This check isn't necessary since the loop won't start if the array is empty.

Reply:
yeah, I don't want to emit on the `_empty` subject if we call closeAll with an empty stack. Seemed cleaner than having a `cont shouldEmit = this._elements.length > 0` variable and checking it later on before calling `next`

Reply:
yeah, but I don't want to emit if the stack was initially empty. Of course I could do something like 
```js
const shouldEmit = this._elements.length > 0;
while(...) {...}
if (shouldEmit) {this._empty.next(focusNext)}
```

But this seemed cleaner IMO